<a href="https://colab.research.google.com/github/BridgingAISocietySummerSchools/Python-for-Data-Science-Workshop/blob/main/01_introduction/04_functions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 🧩 Functions — Reusable Building Blocks

> **Track:** Introduction Session — Notebook 4 of 5 · **Estimated time:** ~12 min · **Difficulty:** complete beginner

The weather report that closed Notebook 3 had its classification rules written out *inline*, in the middle of the loop. That works — until you need the same rule somewhere else. Then you copy, you paste, and you have quietly created a problem for your future self.

This notebook is about the tool that fixes it: the **function** — a named recipe you write **once** and use **as often as you like**.

## 🎯 Learning objectives

By the end of this notebook you will be able to:

1. Explain **why** functions exist, by recognising the copy-paste pain they remove.
2. Define a function with `def`, a parameter list, and an indented body.
3. Call a function several times with different **arguments**.
4. Use `return` to hand a value back — and say precisely how that differs from `print`.
5. Give a parameter a **default value** for convenience.
6. Explain how naming a step makes an analysis readable, testable, and reusable.

## ✅ Prerequisites

Notebooks 1–3: [`01_welcome_to_python.ipynb`](01_welcome_to_python.ipynb) (variables, arithmetic, f-strings), [`02_data_structures.ipynb`](02_data_structures.ipynb) (lists, `len()`) and [`03_control_flow.ipynb`](03_control_flow.ipynb) (`if / elif / else`, `for`, the accumulator pattern). Nothing else is assumed.


## 1. The problem functions solve — copy-paste pain

Let's feel the pain first; the cure will make more sense afterwards.

You are given three weather-station readings in **degrees Fahrenheit** and asked to report them in **degrees Celsius**. With only what you know so far, you would write the conversion once per station — which in practice means writing it once and pasting it twice:


In [1]:
# Three readings in °F, converted to °C — the same arithmetic, copy-pasted.
station_a_c = 68 * 5/9
station_b_c = 95 * 5/9
station_c_c = 41 * 5/9

print(f"Station A: {station_a_c:.1f} °C")
print(f"Station B: {station_b_c:.1f} °C")
print(f"Station C: {station_c_c:.1f} °C")

Station A: 37.8 °C
Station B: 52.8 °C
Station C: 22.8 °C


Three numbers, printed without complaint. Except they are **all wrong**: the Fahrenheit scale does not start at zero, so you must subtract 32 **before** multiplying by 5/9. Python had no way of knowing that — it did exactly the arithmetic we asked for.

So the requirement changes (as it always does) and the formula must be fixed. Note what that means with copy-pasted code: **you must fix it in every copy**. Below, we patch stations A and B, get interrupted by a colleague, and never come back to station C:


In [2]:
station_a_c = (68 - 32) * 5/9    # patched ✅
station_b_c = (95 - 32) * 5/9    # patched ✅
station_c_c = 41 * 5/9           # ...never got patched ❌

print(f"Station A: {station_a_c:.1f} °C")
print(f"Station B: {station_b_c:.1f} °C")
print(f"Station C: {station_c_c:.1f} °C")

print()   # an empty print() just leaves a blank line
print(f"What station C should say: {(41 - 32) * 5/9:.1f} °C")

Station A: 20.0 °C
Station B: 35.0 °C
Station C: 22.8 °C

What station C should say: 5.0 °C


Look at what just happened. Station C reports **22.8 °C** when the true value is **5.0 °C** — a 17.8-degree error in a number that would go straight into a report. No error message, no warning: the code ran perfectly and lied to us, because two lines now follow the new rule and one still follows the old one.

That is not a story about being careless. It is a story about **arithmetic that lives in three places**. With three copies you might catch it; with thirty you will not.

### The cure: write the recipe once, give it a name

A **function** is a named recipe. Think of a machine on a workbench:

```text
        argument                                   returned value
   68  ──────────►  ┌───────────────────────┐  ──────────────────►  20.0
   (input slot)     │  to_celsius(f):       │      (output tray)
                    │    (f - 32) * 5/9     │
                    └───────────────────────┘
```

You build the machine **once**. After that you drop values into the input slot and collect results from the output tray — and if the recipe ever needs fixing, there is exactly **one** place to fix it.

Here is the anatomy, then the real thing:

```text
def to_celsius(fahrenheit):          ← `def`, a name, parameters in (), colon
····return (fahrenheit - 32) * 5/9   ← the body: indented 4 spaces, `return` hands a value back
```

The indentation is Notebook 3's rule again: the colon opens a block, and the indented lines are what belongs inside it.


In [3]:
def to_celsius(fahrenheit):
    return (fahrenheit - 32) * 5/9


# Defining the function printed nothing — we only built the machine.
# Now we USE it, once per station, with a different argument each time:
print(f"Station A: {to_celsius(68):.1f} °C")
print(f"Station B: {to_celsius(95):.1f} °C")
print(f"Station C: {to_celsius(41):.1f} °C")

Station A: 20.0 °C
Station B: 35.0 °C
Station C: 5.0 °C


**20.0, 35.0, 5.0** — correct, all three, including station C. And notice the important part: the formula now appears **exactly once** in this notebook. Station C can no longer disagree with station A, because they are no longer separate copies; they are the same recipe used three times.

The vocabulary, so we can talk about it precisely:

| Term | In our example | Meaning |
|---|---|---|
| `def` | `def to_celsius(...)` | "define a function" — the keyword that starts a recipe |
| **name** | `to_celsius` | how you call it later; name it after *what it produces* |
| **parameter** | `fahrenheit` | the blank on the recipe card, filled in at call time |
| **body** | the indented line | the work, indented 4 spaces under the colon |
| **argument** | the `68` in `to_celsius(68)` | the actual value you hand in |
| **`return`** | `return (fahrenheit - 32) * 5/9` | the value the function hands back |


## 2. Two parameters — and what `return` really buys you

A recipe can have more than one blank. List the parameters inside the brackets, separated by commas, and hand in the arguments in the **same order**.

Here is a function every data analysis needs sooner or later: what percentage is this part of that whole?

In [4]:
def share_percent(part, whole):
    return part / whole * 100


# Two survey groups: 18 of 25 agreed, and 31 of 40 agreed.
print(f"Group A: {share_percent(18, 25):.1f} % agreed")
print(f"Group B: {share_percent(31, 40):.1f} % agreed")

# Because the function RETURNS a value, that value is just a number —
# so we can keep computing with it:
gap = share_percent(31, 40) - share_percent(18, 25)
print(f"Group B leads Group A by {gap:.1f} percentage points")

Group A: 72.0 % agreed
Group B: 77.5 % agreed
Group B leads Group A by 5.5 percentage points


That last line is the whole point of `return` — and mixing up `return` and `print` is *the* most common beginner bug, so let's make it explicit.

## 3. ⚠️ `return` vs `print` — the one distinction to get right

- **`print`** puts characters on your screen. That is *all* it does. The value is gone the moment you have read it.
- **`return`** hands the value back **to your code**, so you can store it in a variable, add it to something, or feed it into another function.

A function that only prints looks fine — you *see* a number, after all — but it secretly returns nothing. Python spells "nothing" as `None`, and `None` cannot be added, averaged, or plotted.

In [5]:
def show_celsius(fahrenheit):
    print((fahrenheit - 32) * 5/9)      # only DISPLAYS the number


def get_celsius(fahrenheit):
    return (fahrenheit - 32) * 5/9      # HANDS THE NUMBER BACK


shown = show_celsius(68)     # the 20.0 below is printed by the function itself
print(f"shown = {shown}")    # → None: the function handed nothing back
print(type(shown))           # type() from Notebook 1 confirms it

got = get_celsius(68)
print(f"got   = {got}")
print(f"got + 5 = {got + 5}")   # a real number, usable in further arithmetic

# Uncomment the next line, run it, and read the error — then comment it out again:
# print(shown + 5)   # TypeError: unsupported operand type(s) for +: 'NoneType' and 'int'

20.0
shown = None
<class 'NoneType'>
got   = 20.0
got + 5 = 25.0


Both functions *displayed* `20.0`, but only one of them **gave** us `20.0`. Rule of thumb: a function that computes something should `return` it; print at the *end*, where you present results — not in the middle, where you are still calculating.

### 🎯 Try it (1–2 min) — Repair a printing function

Your first function of the day. This one is meant to compute a Body Mass Index, but it only prints:

```python
def bmi(weight_kg, height_m):
    print(weight_kg / (height_m * height_m))
```

Rewrite it so it **returns** the value instead, then use it to compute the **average BMI** of two people — 70 kg / 1.75 m and 82 kg / 1.80 m — and print that average with one decimal.


In [6]:
# Your turn 👇  (one line for the def, one for the average, one to print)

# def bmi(weight_kg, height_m):
#     return ...


<details><summary>💡 Click to reveal the solution</summary>

```python
def bmi(weight_kg, height_m):
    return weight_kg / (height_m * height_m)

average_bmi = (bmi(70, 1.75) + bmi(82, 1.80)) / 2
print(f"Average BMI: {average_bmi:.1f}")     # → Average BMI: 24.1
```

**Reasoning.** Three moves, and they are the same three every time: `def` plus a name, a parameter for each thing that varies, and `return` for the answer. Swapping `print` for `return` is a one-word change, but it converts a dead end into a building block: `bmi(70, 1.75)` is now *a number* (22.86), so it can be added, halved, and formatted. With the printing version, `(bmi(...) + bmi(...)) / 2` would have crashed with a `TypeError`, because you cannot add `None` to `None`.
</details>

## 4. A whole list as input — and defaults for convenience

Nothing says an argument has to be a single number. Hand a function a **list** and put a `for` loop in the body, and you have named a whole summarising step: `average(readings)` says what it does, and the loop mechanics disappear inside.

### A parameter can have a default value

Write `rate=0.1` in the parameter list and that parameter becomes **optional**: callers who omit it get the default, callers who care can override it. Same recipe, less typing for the common case.


In [7]:
def average(readings):
    total = 0                      # the accumulator pattern from Notebook 3,
    for value in readings:         # now safely tucked inside a named function
        total = total + value
    return total / len(readings)


week = [18.5, 21.0, 19.5, 25.0, 23.5, 17.0, 20.5]   # °C, one reading per day
print(f"Weekly average: {average(week):.2f} °C")

# The returned value is an ordinary number, so it works anywhere a number does:
if average(week) > 20:
    print("A mild week overall.")


# A default value makes the second argument optional:
def tip(amount, rate=0.1):
    return amount * rate


print(f"Tip on 50 € at the default 10 %: {tip(50):.2f} €")
print(f"Tip on 50 € at a generous 20 %:  {tip(50, 0.2):.2f} €")

Weekly average: 20.71 °C
A mild week overall.
Tip on 50 € at the default 10 %: 5.00 €
Tip on 50 € at a generous 20 %:  10.00 €


`average(week)` reads like a sentence, and `tip(50)` needs no second thought. Both are ordinary functions — nothing new, just applied to a list and given a default.

### 🎯 Try it (1–2 min) — Name a classification step

Write a function `classify(temp)` that takes a temperature in °C and **returns** a label (using `if / elif / else` from Notebook 3):

- `"cold"` if the temperature is below 10,
- `"pleasant"` if it is 25 or below,
- `"hot"` otherwise.

Then loop over the `week` list below and print one line per reading, e.g. `18.5 °C → pleasant`.

In [8]:
week = [18.5, 21.0, 19.5, 25.0, 23.5, 17.0, 20.5]   # still the same readings

# Your turn 👇

<details><summary>💡 Click to reveal the solution</summary>

```python
def classify(temp):
    if temp < 10:
        return "cold"
    elif temp <= 25:
        return "pleasant"
    else:
        return "hot"

for reading in week:
    print(f"{reading} °C → {classify(reading)}")
```

Output:

```text
18.5 °C → pleasant
21.0 °C → pleasant
19.5 °C → pleasant
25.0 °C → pleasant
23.5 °C → pleasant
17.0 °C → pleasant
20.5 °C → pleasant
```

**Reasoning.** Two ideas meet here. First, `return` can appear **more than once**: the first `return` that runs ends the function immediately, which is why `if / elif / else` and `return` are such a natural pair. Second, the loop no longer contains any *rules* — only `classify(reading)`. The decision lives in one named place, so changing where "pleasant" ends is a one-line edit. (This mild week is all `"pleasant"`; add a `2.0` or a `31.0` to the list and re-run to see the other branches fire.)
</details>

## 5. Why functions structure code

Removing duplication was only the beginning. The deeper gain is that **naming a step turns a wall of code into a description of your analysis**. `average(week)`, `share_percent(18, 25)`, `classify(reading)` — a reader can follow the argument without decoding a single line of arithmetic. That buys three things at once:

- **Readable.** The name says the intent; the mechanics stay out of the way until you need them.
- **Testable.** You can check `to_celsius(212)` and expect exactly `100.0`. A formula smeared across forty lines has nothing you can point at and check.
- **Reusable.** Written once, correct once. The station-C bug from Section 1 is not merely fixed — it has become *impossible*.

And this is not a beginner's habit you will outgrow. It is how the whole ecosystem is built: every `pd.read_csv(...)` and `.describe()` you will write in Notebook 5 is **someone else's function** — a named recipe, with parameters, that hands a value back. Learning to write your own is learning to read theirs.

### 💬 Discuss (30 seconds, with your neighbour)

Our `classify()` function hides a decision: *someone chose* that "pleasant" ends at 25 °C. For weather, who cares. But the same shape of code decides whether an applicant is "qualified", a transaction is "suspicious", or a patient is "high risk" — and the threshold sits quietly inside the function, invisible to everyone reading the output.

**When a function makes a judgement about people, who should be able to look inside it — and how would they even know a threshold was there?** Does it change your answer if the numbers inside were not chosen by a person at all, but learned from data?


## 🧠 Key takeaways

1. **Copy-pasted logic rots.** When a rule changes you must fix every copy, and the copy you miss produces wrong numbers *silently* — no error, just a report that disagrees with itself.
2. **A function is a named recipe**: `def name(parameters):` followed by an indented body. Defining it does nothing visible; **calling** it does the work.
3. **Parameters are the blanks; arguments are what you put in them** — one, two, or a whole list.
4. **`return` hands a value back to your code; `print` only shows it.** A function that just prints returns `None`, and `None` cannot be added, averaged, or plotted. Compute with `return`, present with `print`.
5. **Default values** (`def tip(amount, rate=0.1)`) make a parameter optional — convenience for the common case, flexibility when you need it.
6. **Naming a step makes an analysis readable, testable, and reusable** — and every library call you will ever make is somebody else's function.

## ✅ Self-assessment

Before moving on, check that you can:

- [ ] Explain in one sentence why three copies of a formula are a liability
- [ ] Write a function with `def`, a parameter, an indented body, and a `return`
- [ ] Call your function several times with different arguments
- [ ] Use a function's returned value in further arithmetic
- [ ] Say what a function returns when it only prints — and why that matters
- [ ] Read `def tip(amount, rate=0.1)` and explain what `tip(50)` does
- [ ] Put a `for` loop inside a function and return the summarised result

## 🚀 Next step

You now have the building blocks of Python: **data** (Notebooks 1–2), **decisions and repetition** (Notebook 3), and **functions** (this one). Time to see what professionals actually do with them.

Continue with **Notebook 5 — First Steps with Pandas** ([`05_pandas_intro.ipynb`](05_pandas_intro.ipynb)): a whole table becomes a single object, and one call to somebody else's function does what a hand-written loop needed ten lines for.

> **Want to go deeper after the session?** The self-paced track in [`../02_advanced_self_learning/`](../02_advanced_self_learning/) revisits functions in full depth — [`05_functions_and_modules.ipynb`](../02_advanced_self_learning/05_functions_and_modules.ipynb) adds scope, `*args`, lambdas and your own modules.
